runtime/src/util/events.rs

```diff
@@ -0,0 +1,127 @@
+use crate::cbor::serialize_vec;
+use crate::ActorError;
+use fvm_shared::event::{ActorEvent, Entry, Flags};
+use serde::ser;
+
+// Codec identifier for CBOR-encoded data.
+const IPLD_CBOR: u64 = 0x51;
+
+const EVENT_TYPE_KEY: &str = "$type";
+
+/// Builder for ActorEvent objects, accumulating key/value pairs.
+pub struct EventBuilder {
+    entries: Result<Vec<Entry>, ActorError>,
+}
+
+impl EventBuilder {
+    /// Creates a new builder with no values.
+    pub fn new() -> Self {
+        Self { entries: Ok(Vec::new()) }
+    }
+
+    /// Initialise the "type" of the event i.e. Actor event type.
+    pub fn event_type(self, _type: &str) -> Self {
+        self.push_entry(EVENT_TYPE_KEY, _type, Flags::FLAG_INDEXED_KEY)
+    }
+
+    /// Pushes an entry with an indexed key and an un-indexed, IPLD-CBOR-serialized value.
+    pub fn field<T: ser::Serialize + ?Sized>(self, name: &str, value: &T) -> Self {
+        self.push_entry(name, value, Flags::FLAG_INDEXED_KEY)
+    }
+
+    /// Pushes an entry with an indexed key and indexed, IPLD-CBOR-serialized value.
+    pub fn field_indexed<T: ser::Serialize + ?Sized>(self, name: &str, value: &T) -> Self {
+        self.push_entry(name, value, Flags::FLAG_INDEXED_ALL)
+    }
+
+    /// Returns an actor event ready to emit (consuming self).
+    pub fn build(self) -> Result<ActorEvent, ActorError> {
+        Ok(ActorEvent { entries: self.entries? })
+    }
+
+    /// Pushes an entry with an IPLD-CBOR-serialized value.
+    fn push_entry<T: ser::Serialize + ?Sized>(
+        mut self,
+        key: &str,
+        value: &T,
+        flags: Flags,
+    ) -> Self {
+        if let Ok(ref mut entries) = self.entries {
+            match serialize_vec(&value, "event value") {
+                Ok(value) => {
+                    entries.push(Entry { flags, key: key.to_string(), codec: IPLD_CBOR, value })
+                }
+                Err(e) => {
+                    self.entries = Err(e);
+                }
+            }
+        }
+        self
+    }
+}
+
+impl Default for EventBuilder {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::cbor::serialize_vec;
+    use crate::util::events::{EVENT_TYPE_KEY, IPLD_CBOR};
+    use crate::EventBuilder;
+    use fvm_shared::event::{ActorEvent, Entry, Flags};
+
+    #[test]
+    fn event_type() {
+        let e = EventBuilder::new().event_type("l1").event_type("l2").build().unwrap();
```

Comment:
Two types on one event would be pretty uncommon, so it's odd as a test case.

Reply:
Done with a better test.